services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -380,4 +283,286 @@ private ServerReflectionResponse createServerReflectionResponse(
+          // Service names may change without a change in the underlying file descriptors if one
+          // proto file defines multiple services.
+          mutableServicesIndex.setServiceNames(currentMutableServiceNames);
+        }
+      }
+    }
+
+    private Set<String> getServiceNames() {
+      Set<String> serviceNames = new HashSet<String>(immutableServicesIndex.getServiceNames());
+      synchronized (mutableServicesIndex) {
+        if (mutableServicesIndex.isInitialized()) {
+          serviceNames.addAll(mutableServicesIndex.getServiceNames());
+        }
+      }
+      return serviceNames;
+    }
+
+    @Nullable
+    private FileDescriptor getFileDescriptorByName(String name) {
+      FileDescriptor fd = immutableServicesIndex.getFileDescriptorByName(name);
+      if (fd == null) {
+        synchronized (mutableServicesIndex) {
+          if (mutableServicesIndex.isInitialized()) {
+            fd = mutableServicesIndex.getFileDescriptorByName(name);
+          }
+        }
+      }
+      return fd;
+    }
+
+    @Nullable
+    private FileDescriptor getFileDescriptorBySymbol(String symbol) {
+      FileDescriptor fd = immutableServicesIndex.getFileDescriptorBySymbol(symbol);
+      if (fd == null) {
+        synchronized (mutableServicesIndex) {
+          if (mutableServicesIndex.isInitialized()) {
+            fd = mutableServicesIndex.getFileDescriptorBySymbol(symbol);
+          }
+        }
+      }
+      return fd;
+    }
+
+    @Nullable
+    private FileDescriptor getFileDescriptorByExtensionAndNumber(String type, int extension) {
+      FileDescriptor fd =
+          immutableServicesIndex.getFileDescriptorByExtensionAndNumber(type, extension);
+      if (fd == null) {
+        synchronized (mutableServicesIndex) {
+          if (mutableServicesIndex.isInitialized()) {
+            fd = mutableServicesIndex.getFileDescriptorByExtensionAndNumber(type,
+                extension);
+          }
+        }
+      }
+      return fd;
+    }
+
+    @Nullable
+    private Set<Integer> getExtensionNumbersOfType(String type) {
+      Set<Integer> extensionNumbers = immutableServicesIndex.getExtensionNumbersOfType(type);
+      if (extensionNumbers == null) {
+        synchronized (mutableServicesIndex) {
+          if (mutableServicesIndex.isInitialized()) {
+            extensionNumbers = mutableServicesIndex.getExtensionNumbersOfType(type);
+          }
+        }
+      }
+      return extensionNumbers;
+    }
+
+    /**
+     * Provides a set of methods for answering reflection queries for the file descriptors
+     * underlying a set of services. Used by {@link ServerReflectionIndex} to separately index
+     * immutable and mutable services.
+     *
+     * <p>This class is not thread-safe. {@link ServerReflectionIndex} handles synchronization for
+     * initialization, updates (for mutable services), and reads.
+     */
+    private class FileDescriptorIndex {
```

Comment:
static final, if possible.

It is extremely hard to change this later, since there are tests that check that the signatures of their libraries don't change.  (even though it is absurd to check).